src/pull.rs

```diff
@@ -0,0 +1,234 @@
+use crate::meta_store::MetaStore;
+use crate::unpack::unpack;
+
+const DIGEST_SHA256: &str = "sha256";
+const DIGEST_SHA512: &str = "sha512";
+
+/// The PullClient connects to remote OCI registry, pulls the container image,
+/// and save the image layers under data_dir and return the layer meta info.
+pub struct PullClient {
+    /// `oci-distribuion` client to talk with remote OCI registry.
+    pub client: Client,
+
+    /// OCI registry auth info.
+    pub auth: RegistryAuth,
+
+    /// OCI image reference.
+    pub reference: Reference,
+
+    /// OCI image layer data store dir.
+    pub data_dir: PathBuf,
+}
+
+impl PullClient {
+    /// Constructs a new PullClient struct with provided image info,
+    /// data store dir and optional remote registry auth info.
+    pub fn new(image: &str, data_dir: &Path, auth_info: &Option<&str>) -> Result<PullClient> {
+        let mut auth = RegistryAuth::Anonymous;
+        if let Some(auth_info) = auth_info {
+            if let Some((username, password)) = auth_info.split_once(':') {
+                auth = RegistryAuth::Basic(username.to_string(), password.to_string());
+            } else {
+                return Err(anyhow!("Invalid authentication info ({:?})", auth_info));
+            }
+        }
+
+        let reference = Reference::try_from(image)?;
+        let client = Client::default();
+
+        Ok(PullClient {
+            client,
+            auth,
+            reference,
+            data_dir: data_dir.to_path_buf(),
+        })
+    }
+
+    /// pull_manifest pulls an image manifest and config data.
+    pub async fn pull_manifest(&mut self) -> Result<(OciManifest, String, String)> {
+        Ok(self
+            .client
+            .pull_manifest_and_config(&self.reference, &self.auth)
+            .await?)
+    }
+
+    /// pull_layers pulls an image layers and do ondemand decrypt/decompress.
+    /// It returns the layer metadata for layer db to track.
+    pub async fn pull_layers(
+        &self,
+        layer_descs: Vec<OciDescriptor>,
+        diff_ids: &[String],
+        decrypt_config: &Option<&str>,
+        meta_store: Arc<Mutex<MetaStore>>,
+    ) -> Result<Vec<LayerMeta>> {
+        let layer_metas = layer_descs.into_iter().enumerate().map(|(i, layer)| {
+            let client = &self.client;
+            let reference = &self.reference;
+            let ms = meta_store.clone();
+            async move {
+                let mut out: Vec<u8> = Vec::new();
+                let mut layer_data: Vec<u8> = Vec::new();
+                let plaintxt_layer: Vec<u8>;
+
+                client
+                    .pull_layer(reference, &layer.digest, &mut layer_data)
+                    .await?;
+
+                let mut layer_meta = LayerMeta::default();
+                let mut media_type_str: &str = layer.media_type.as_str();
+
+                let mut decryptor = Decryptor::default();
+                if decryptor.is_encrypted(&layer.media_type) {
```

Comment:
One of the issues we ran into with `skopeo` was that if a decryption key was provided but the image was not encrypted, the image would be unpacked anyway. For CCv0 this meant that if the CSP intercepted an encrypted image and replaced it with an unencrypted image, `skopeo` would unpack the unencrypted image inside the enclave without warning the user.

Does `image-rs` do something similar? Here it looks like the format of the image determines whether the image is decrypted or not. Do we check whether or not encryption is expected from by the client?

Reply:
Yes, current PR for `image-rs` do the same way as `skopeo`, and after we pulled the image, we will return the image and layer metadata info like whether the image is signed and layer is encrypted. 

The next step work for integrating `signature` crate which also provide the policy info, we will do these verifications to ensure the image/layer info is comply with the policy.

Reply:
Ok, let's just make sure we enforce the guest owner's policy at some point, even for unsigned images.